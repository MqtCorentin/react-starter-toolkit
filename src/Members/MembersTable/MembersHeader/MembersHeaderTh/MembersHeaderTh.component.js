import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { ASCENDING, DESCENDING, NONE } from 'shared/constants';

export const MembersHeaderThSortable = ({ children, sort, order, sortable, sortingIcon }) => (
  <th
    role="button"
    onClick={sort}
    className={classNames('af-table__th', {
      'af-table__th--sortable': Boolean(sortable),
      'af-table__th--active': Boolean(order !== NONE),
    })}
  >
    <button
      type="button"
      className={classNames('af-btn', {
        'af-btn--table-sorting': Boolean(sortable),
        'af-btn--table-sorting--active': Boolean(order !== NONE),
      })}
    >
      {children}
      {sortingIcon}
    </button>
  </th>
);

const MembersHeaderThSortablePropTypes = {
  children: PropTypes.node,
  sort: PropTypes.func,
  order: PropTypes.oneOf([NONE, ASCENDING, DESCENDING]),
  sortable: PropTypes.bool.isRequired,
};

const MembersHeaderThSortableDefaultProps = {
  order: NONE,
  sort: undefined,
  children: null,
};

MembersHeaderThSortable.propTypes = MembersHeaderThSortablePropTypes;
MembersHeaderThSortable.defaultProps = MembersHeaderThSortableDefaultProps;

export const MembersHeaderTh = ({ children }) => (
  <th className="af-table__th" role="columnheader">
    {children}
  </th>
);
