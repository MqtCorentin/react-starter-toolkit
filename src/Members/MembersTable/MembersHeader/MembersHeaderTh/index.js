export { default } from './MembersHeaderTh.container';
