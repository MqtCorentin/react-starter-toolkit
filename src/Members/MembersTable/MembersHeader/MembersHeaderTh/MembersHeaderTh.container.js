//
import React from 'react';
import { ASCENDING, DESCENDING, NONE } from 'shared/constants';
import { MembersHeaderThSortable, MembersHeaderTh } from './MembersHeaderTh.component';

export const toggleOrder = order => (order === ASCENDING ? DESCENDING : ASCENDING);

export const toggleSorting = (field, { field: prevField, order }) => {
  const newOrder = {
    field,
    order: prevField !== field ? ASCENDING : toggleOrder(order),
  };
  return newOrder;
};

const orderIcons = (order = NONE) =>
  ({
    [NONE]: 'sorting',
    [ASCENDING]: 'arrow-xs-up',
    [DESCENDING]: 'arrow-xs-down',
  }[order]);

export const MembersHeaderThEnhanced = ({ sortingInfo, field, onSort, sorting, ...rest }) => {
  const sortable = Boolean(field);
  const order = sortingInfo[field];

  const computedProps = {
    sortable,
    order,
    sort: () => onSort(toggleSorting(field, sorting)),
    sortingIcon: sortable ? <span className={`af-btn__icon af-btn__icon--table-sorting glyphicon glyphicon-${orderIcons(order)}`} /> : null,
  };

  return sortable ? <MembersHeaderThSortable {...rest} {...computedProps} /> : <MembersHeaderTh {...rest} />;
};

export default MembersHeaderThEnhanced;
