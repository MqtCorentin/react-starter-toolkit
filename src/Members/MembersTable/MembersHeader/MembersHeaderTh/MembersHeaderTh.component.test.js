import React from 'react';
import { render } from '@testing-library/react';
import { MembersHeaderThSortable, MembersHeaderTh } from './MembersHeaderTh.component';

describe('<MembersHeaderTh/>', () => {
  it('Should render MembersHeaderTh with Hello text', () => {
    const tr = document.createElement('tr');
    const { asFragment, getByText } = render(<MembersHeaderTh>Hello</MembersHeaderTh>, {
      container: document.body.appendChild(tr),
    });
    expect(getByText('Hello')).toBeDefined();
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('<MembersHeaderThSortable />', () => {
  it('Should render MembersHeaderThSortable with Hello text', () => {
    const tr = document.createElement('tr');
    const { asFragment, getByText, getByRole } = render(
      <MembersHeaderThSortable sortable order={1} sortingIcon={<span role="img">♛</span>}>
        Hello
      </MembersHeaderThSortable>,
      {
        container: document.body.appendChild(tr),
      },
    );
    expect(getByText('Hello')).toBeDefined();
    expect(getByText('♛')).toBeDefined();
    expect(getByRole('img')).toBeDefined();
    expect(asFragment()).toMatchSnapshot();
  });
});
