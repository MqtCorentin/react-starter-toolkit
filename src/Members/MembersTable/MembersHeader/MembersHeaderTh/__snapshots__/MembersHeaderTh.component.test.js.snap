// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MembersHeaderTh/> Should render MembersHeaderTh with Hello text 1`] = `
<DocumentFragment>
  <th
    class="af-table__th"
    role="columnheader"
  >
    Hello
  </th>
</DocumentFragment>
`;

exports[`<MembersHeaderThSortable /> Should render MembersHeaderThSortable with Hello text 1`] = `
<DocumentFragment>
  <th
    class="af-table__th af-table__th--sortable af-table__th--active"
    role="button"
  >
    <button
      class="af-btn af-btn--table-sorting af-btn--table-sorting--active"
      type="button"
    >
      Hello
      <span
        role="img"
      >
        â™›
      </span>
    </button>
  </th>
</DocumentFragment>
`;
